[LLDB][ELF] Fix section unification to not just use names.

al45tair opened the pull request:
Section unification cannot just use names, because it's valid for ELF binaries to have multiple sections with the same name.  We should check other section properties too.

Fixes #88001.

rdar://124467787

al45tair committed:
[LLDB][ELF] Fix section unification to not just use names.

Section unification cannot just use names, because it's valid for ELF
binaries to have multiple sections with the same name.  We should check
other section properties too.

rdar://124467787

JDevlieghere commented on lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp:
```diff
@@ -1854,6 +1854,49 @@ class VMAddressProvider {
 };
 }
 
+namespace {
```
Nit: LLVM and LLDB prefer static functions over anonymous namespaces.

al45tair replied:
This file already uses anonymous namespaces, but I'll change the function to `static` instead.

JDevlieghere commented on lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp:
```diff
@@ -1854,6 +1854,49 @@ class VMAddressProvider {
+         sect_iter != section_list.end();
+         ++sect_iter) {
+      if ((*sect_iter)->GetName() == name
+          && (*sect_iter)->GetType() == type
+          && (*sect_iter)->IsThreadSpecific() == thread_specific
+          && (*sect_iter)->GetPermissions() == permissions
+          && (*sect_iter)->GetFileSize() == file_size
+          && (*sect_iter)->GetByteSize() == byte_size
+          && (*sect_iter)->GetFileAddress() == vm_addr
+          && (*sect_iter)->GetLog2Align() == alignment) {
```
This seems like it could be an `operator ==` in the Section class.

al45tair replied:
I didn't want to do that, because I didn't want to give the impression that this was a general purpose way to compare two `Section`s. While it checks _many_ of the section attributes, it doesn't check all of them (intentionally in the case of which object they come from), and I think that would be surprising behaviour from an `==` operator.

labath replied:
Am I correct in understanding that this code is used when unifying the sections of a separate debug file with a stripped file? If so, then I believe this comparison is too strict. An object file produced by `objcopy --only-keep-debug` will only have placeholder sections instead of the sections that contained actual code. That means (at least) their file size and file offset will be different from that in the original file.

I think it would be sufficient to use the file address (called virtual address in ELF), possibly confirmed by section name, as the unification key.

al45tair replied:
I did wonder about exactly how strict we want to be (and exactly which things to compare). This isn't testing file offset, but you're right that it does check file size and maybe it shouldn't.

al45tair replied:
@labath Let's get rid of the sizes; the other things should all be the same, correct?

labath replied:
The section type can also change. You can see that with .text, because there we have a name-based fallback, but with anything else, it changes from "code" to "regular":
```$ bin/lldb-test object-file /tmp/a.debug  | grep -e foo -C 3
  Showing 1 subsections
    Index: 0
    ID: 0x6
    Name: .foo
    Type: regular
    Permissions: r-x
    Thread specific: no
$ bin/lldb-test object-file /tmp/a.out  | grep -e foo -C 3
  Showing 1 subsections
    Index: 0
    ID: 0x6
    Name: .foo
    Type: code
    Permissions: r-x
    Thread specific: no
```

The other fields are _probably_ fine.

al45tair replied:
OK. I've removed the section type test as well.

JDevlieghere commented on lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp:
```diff
@@ -1854,6 +1854,49 @@ class VMAddressProvider {
+    uint32_t permissions = section->GetPermissions();
+    uint32_t alignment = section->GetLog2Align();
+
+    for (auto sect_iter = section_list.begin();
+         sect_iter != section_list.end();
+         ++sect_iter) {
```
Can this be a range-based for loop?

al45tair replied:
Yes, it can (I just checked and apparently we can use C++17 in LLVM code, so that's nice :-)).

JDevlieghere commented on lldb/source/Plugins/ObjectFile/ELF/ObjectFileELF.cpp:
```diff
@@ -2067,10 +2110,8 @@ unsigned ObjectFileELF::ParseSymbols(Symtab *symtab, user_id_t start_id,
-  // char*" key must came from a ConstString object so they can be compared by
-  // pointer
-  std::unordered_map<const char *, lldb::SectionSP> section_name_to_section;
+  // Cache the section mapping
```
Nit: missing period. A map from SectionSP to SectionSP looks somewhat suspicious, might be worth explaining what the key and the value mean.

al45tair replied:
I'll improve the comment.

JDevlieghere reviewed (commented):
Can we add a test for this with obj2yaml?

al45tair commented:
> Can we add a test for this with obj2yaml?

Not sure what you're thinking here. We can't use `yaml2obj` to generate an object with this problem because that tool assumes symbols are in a uniquely named section (which is mostly reasonable), and `obj2yaml` doesn't exercise the code in LLDB. I'll take a look and see what we can do as a test.

al45tair committed:
[LLDB][ELF] Address review feedback, add test.

Fixed a couple of nits from review, and fixed up formatting.

Also added a test.

rdar://124467787

labath commented:
> > Can we add a test for this with obj2yaml?
> 
> Not sure what you're thinking here. We can't use `yaml2obj` to generate an object with this problem because that tool assumes symbols are in a uniquely named section 

[this](https://github.com/llvm/llvm-project/blob/main/llvm/test/tools/yaml2obj/ELF/duplicate-section-names.yaml) test seems to indicate that's possible (the trick appears to me in `(n)` name tag suffixes). Can you give that a shot?

al45tair commented:
> [this](https://github.com/llvm/llvm-project/blob/main/llvm/test/tools/yaml2obj/ELF/duplicate-section-names.yaml) test seems to indicate that's possible (the trick appears to me in `(n)` name tag suffixes). Can you give that a shot?

I've actually just included a binary file that exhibits the problem.

al45tair committed:
[LLDB][ELF] Don't match on file size.

As pointed out by @labath, if you use `objcopy --only-keep-debug`, only
placeholder sections will be present so the file sizes won't match.

We already ignore file offsets when doing the comparison.

rdar://124467787

labath commented:
> > [this](https://github.com/llvm/llvm-project/blob/main/llvm/test/tools/yaml2obj/ELF/duplicate-section-names.yaml) test seems to indicate that's possible (the trick appears to me in `(n)` name tag suffixes). Can you give that a shot?
> 
> I've actually just included a binary file that exhibits the problem.

I saw that, but a textual test is definitely preferable, particularly after the linux xz fiasco. This wouldn't be the first test binary in our repo, but in this case, I think it actually adds a lot of value, since yaml2obj operates on the same level as what you are testing (so you can see the input of the test eyeball-verify that the expected output makes sense).

al45tair committed:
[LLDB][ELF] Use `yaml2obj` for the test.

Rather than including a binary, use `yaml2obj` for the duplicate section
name test.

rdar://124467787

labath commented:
> > I saw that, but a textual test is definitely preferable, particularly after the linux xz fiasco. This wouldn't be the first test binary in our repo, but in this case, I think it actually adds a lot of value, since yaml2obj operates on the same level as what you are testing (so you can see the input of the test eyeball-verify that the expected output makes sense).
> 
> :-) Fair point (though this is very much not like the xz incident — I'm a colleague of Jonas's at Apple, albeit in a different team, with a fairly longstanding history of contributing to OSS projects, and doing something like that would be… very career limiting).
> 
> I'll have a go at changing things to use `yaml2obj`.

The new test looks great.

al45tair committed:
[LLDB][ELF] Also ignore section type when matching.

The section type can change when using `objcopy --only-keep-debug`,
apparently.

rdar://124467787

al45tair commented:
@labath @JDevlieghere Are we happy with this PR now? If so, I'll cherry pick it to the various Apple forks; I need it in order to merge https://github.com/apple/swift/pull/72061, which we need for both my fully statically linked Swift SDK work and for Amazon Linux 2023 support in Swift.

JDevlieghere reviewed (approved):
Looks good to me as well.
